src/libs/elektra/keyname.c

```diff
@@ -255,10 +301,11 @@ keyDel(key);
  */
 const char * keyName (const Key * key)
 {
-	if (!key) return 0;
+	if (!key || !key->keyName) return 0;
 
-	ELEKTRA_ASSERT (key->key != NULL, "invalid name");
-	return key->key;
+	ELEKTRA_ASSERT (key->keyName->key != NULL, "invalid name");
```

Comment:
To match unescaped name below:

```suggestion
	if (!key) return 0;
	ELEKTRA_ASSERT (key->keyName != NULL, "invalid name");
	ELEKTRA_ASSERT (key->keyName->key != NULL, "invalid name");
```

Reply:
Hmm, based on the docs: `@retval "" when Key's name is empty`.

So wouldn't it be better if we return an empty string if `key->keyName` is `NULL`?

Reply:
If you are talking about the return values of `keyName()`: Please be very faithful to the public API **exactly** as it is. The API was recently completely reviewed. Changes must be very careful, usually even with decisions and also all changes completely listed in the release notes (for later 1.0 migration documents). So it is not a good idea to do such changes within other PRs.

What `key->keyName` stores is more or less an implementation detail. It is actually going to change, together with an API change of `keyName` which probably will solve anyway what you ask here? See doc/decisions/2_in_progress/store_name.md

Reply:
> Please be very faithful to the public API exactly as it is. 

Yes, I'm trying. but the current implementation (meaning the implementation before COW) also is not faithful. It only has two cases:

1. The `key` is `NULL` -> returns 0
2. Everything else -> returns what is stored as unescaped name, event if it is `NULL`.

It does never explicitly check whether the name itself is `NULL` to return an empty string.

Reply:
I have updated `keyName` and `keyUnescapedName` now to conform to the API description

Reply:
Okay this creates too much trouble. The implicit assumption in many function is, that a `NULL` keyname returns `NULL`. Reverting it back.

Reply:
A bit of explanation: There was no `if` check for `key->key == NULL`, because that can never happen, when a key is created properly via the API. However, adding an assert just in case would make sense. The same question probably transfers to the new COW struct.

There was no `return ""` and there should never be one, because that is **not** a valid keyname.

Reply:
Yes this can actuall happen! 

`ksLookupByName` creates its own key without the public API. If the `name` passed in is not a valid key name, it will be `NULL` as the return value of `keySetName` is never checked ....

https://github.com/ElektraInitiative/libelektra/blob/e9f2cab6f743f63746bfb800a3b4d14a29fe7df6/src/libs/elektra/keyset.c#L2608-L2627

Reply:
Well this was apparently the root of the problem. Fixed this, now everything works again.

Reply:
> `ksLookupByName` creates its own key without the public API

Then this is a problem in `ksLookupName`. In `keyName` we only need to expect (`if` check) what is possible via the public API, but we should still guard against (`assert`) other issues.

In the specific case, `ksLookupByName` should return `NULL`, when `keySetName` fails. Anything else is simply a bug.